src/conversions/num_complex.rs

```diff
@@ -152,6 +152,19 @@ macro_rules! complex_conversion {
 
                 #[cfg(any(Py_LIMITED_API, PyPy))]
                 unsafe {
+                    use crate::{intern, type_object::PyTypeInfo};
+
+                    let obj = if PyComplex::is_type_of(obj) {
+                        obj
+                    } else if let Ok(method) =
+                        obj.get_type().getattr(intern!(obj.py(), "__complex__"))
```

Comment:
Why `.get_type().getattr(..)` instead of just `.getattr(..)`?

Reply:
[The spec says magic methods are accessed off the type](https://docs.python.org/3/reference/datamodel.html#special-method-names), so I was trying to be consistent with this behaviour:
```python
class A:
    def __complex__(self):
        return 1j
a = A()
a.__complex__ = lambda self: 2j
print(complex(a))  # 1j
```

In saying this, though, I realise that I'm not invoking the descriptor protocol on access to the `__complex__` attribute of the type, if required, which is why I guess the word "roughly" is in:
> `x[i]` is roughly equivalent to `type(x).__getitem__(x, i)`

Does PyO3 have a convenient interface for manually invoking the descriptor protocol on an object? CPython does it very manually in its internal `_PyObject_LookupSpecial` method, which is what it typically uses to access `__complex__`.

Reply:
I'd thought about this, but I was concerned about the additional logic in that for parsing `str` and subclasses.  I can swap it to that if you think it's preferable, likely with an catch to error out if we're given a string?  The current implementation is pretty similar to what CPython does (except for me forgetting to invoke the descriptor protocol right now), but I get that over-tight coupling between PyO3/CPython isn't ideal.

Reply:
Yeah, this too much complexity, especially if we try to second guess it by avoid the string logic.

I am also not sure whether `.get_type().getattr(..)` only lacks the descriptor protocol compared to `_PyObject_LookupSpecial` which also seems to do MRO look-up via `_PyType_Lookup`?

To be honest, I wonder if the the current implementation is good enough and a comment on this being "almost" the same protocol as the one CPython uses.

Reply:
I think the MRO lookup functions correctly with `obj.get_type().getattr()` - the result of `obj.get_type()` is just a regular `PyTypeObject` so the subsequent `getattr` call should resolve the MRO.  I put in an inheritance test in aa1dcb0461ac428d0dc245fe4d3ad1b5e1322028, I put in the implementation of the descriptor protocol (unfortunately the `PyTypeObject` struct is non-limited API, so no `tp_descr_get` slot for us here) in a599e82e, but if you think that's too complex, I can just remove that commit.

Reply:
I think the code is fine and the tests are convincing.

I would like to avoid sinking this into this `extract` implementation. We should factor out a crate-wide (but internal) helper method, something like `lookup_special` implementing the whole `.get_type().attr(..)`-and-then-again dance starting from a name like `__complex__`.

Maybe as a method on `PyAny` that we could eventually (but not in this PR) make public?

> unfortunately the PyTypeObject struct is non-limited API, so no tp_descr_get slot for us here

I think this is fine for this use case as we already are in a lower performance fallback. For the general helper method, it would be nice to add both implementations (limited and non-limited API). But I do not want to steal your time as this is tangential to the issue you are working on, so a FIXME comment would work for me as well.

Reply:
I'm happy to do that too, no worries (but thanks for the consideration!).  Would you prefer a separate clean PR for that implementation, then I can update this PR to use it after that's merged?

I'm not in any hurry on this PR so I've got time to follow whatever procedures you prefer.  I found the `__complex__` discrepancy because of a bug in our own code (we didn't mean to use the behaviour, but were incorrectly returning a single-valued Numpy array at a point we expected a scalar).  We fixed our bug, so we could immediately move to `abi3` regardless of this PR.

Reply:
> Would you prefer a separate clean PR for that implementation, then I can update this PR to use it after that's merged?

I think it would be fine to do here as the PR is currently very self-contained and it is always nice to merge infrastructure together with at least one usage site.